src/parser/validator.cc

```diff
@@ -0,0 +1,218 @@
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include "maliput_sparse/parser/validator.h"
+
+#include "maliput_sparse/geometry/utility/geometry.h"
+#include "maliput_sparse/parser/junction.h"
+#include "maliput_sparse/parser/lane.h"
+#include "maliput_sparse/parser/segment.h"
+
+namespace maliput_sparse {
+namespace parser {
+
+namespace {
+
+// Convenient constant for indicating left or right.
+static constexpr bool kLeft{true};
+static constexpr bool kRight{!kLeft};
+
+}  // namespace
+
+Validator::Validator(const Parser* parser, const ValidatorOptions& options, const ValidatorConfig config) {
+  MALIPUT_THROW_UNLESS(parser);
+  if (options.lane_adjacency) {
+    ValidateLaneAdjacency(parser, config);
+  }
+}
+
+const std::vector<Validator::Error>& Validator::GetErrors() const { return errors_; }
+
+void Validator::CheckAdjacency(const Lane& lane, const Lane& adjacent_lane, bool left, double tolerance) {
+  const auto line_string_a = left ? lane.left : lane.right;
+  const auto line_string_b = left ? adjacent_lane.right : adjacent_lane.left;
+  const double distance = geometry::utility::ComputeDistance(line_string_a, line_string_b);
+  if (distance > tolerance) {
+    const std::string msg{"Lane " + lane.id + " and lane " + adjacent_lane.id +
+                          " are not adjacent under the tolerance " + std::to_string(tolerance) +
+                          ". (distance: " + std::to_string(distance) + ")"};
+    Report(msg, Validator::Error::Type::kLaneAdjacency, Validator::Error::Severity::kError);
+  }
+}
+
+void Validator::Report(const std::string& message, const Validator::Error::Type& type,
+                       const Validator::Error::Severity& severity) {
+  errors_.push_back(Validator::Error{message, type, severity});
+}
+
+void Validator::ValidateLaneAdjacency(const Parser* parser, const ValidatorConfig config) {
```

Comment:
What if this function returns a vector of Errors? Wouldn't it make Validator just a collector of errors by calling Validate*(). You can then collect them in the for at the construction or the operator() overload.

Reply:
I transformed the class into a functor, in a way that the validation occurs when `operator()` is called.